File: mobile/android/app/src/main/java/com/bytefolk/roleweave/PhoneLinkCodec.kt
package com.bytefolk.roleweave

import org.json.JSONObject

object PhoneLinkCodec {
    fun parseSnapshot(raw: String): Snapshot {
        val json = JSONObject(raw)
        val roles = json.getJSONArray("roles")
        val list = buildList {
            for (i in 0 until roles.length()) {
                val item = roles.getJSONObject(i)
                add(
                    Role(
                        id = item.getString("id"),
                        name = item.getString("name"),
                        description = item.optString("description"),
                        skillExcerpt = item.optString("skillExcerpt"),
                    )
                )
            }
        }
        return Snapshot(
            name = json.getString("name"),
            description = json.optString("description"),
            roles = list,
        )
    }

    fun helloPayload(deviceToken: String): String =
        JSONObject()
            .put("v", 1)
            .put("type", "phone.hello")
            .put("deviceToken", deviceToken)
            .toString()

    fun commandPayload(commandId: String, text: String, positionId: String?): String {
        val payload = JSONObject()
            .put("v", 1)
            .put("type", "command.submit")
            .put("commandId", commandId)
            .put("text", text)
        if (!positionId.isNullOrEmpty()) payload.put("positionId", positionId)
        return payload.toString()
    }

    /** Exponential backoff: 1s, 2s, 4s, 8s, 16s for attempts 1..5. */
    fun reconnectDelayMs(attempt: Int): Long {
        val n = attempt.coerceAtLeast(1).coerceAtMost(16)
        return 1000L shl (n - 1)
    }

    fun commandStatusLabel(state: String): String =
        when (state) {
            "accepted" -> "电脑已接到"
            "running" -> "员工正在处理"
            "completed" -> "完成"
            "failed" -> "失败"
            "busy" -> "该员工正在忙"
            "needs_approval" -> "请在电脑上确认"
            else -> state
        }
}

File: mobile/ios/RoleWeave/PhoneLinkCodec.swift
import Foundation

enum PhoneLinkCodec {
    static let maxReconnectAttempts = 5

    static func reconnectDelayMs(attempt: Int) -> Int {
        let clamped = max(1, min(attempt, maxReconnectAttempts))
        return 1000 * (1 << (clamped - 1))
    }

    static func isValidPairCode(_ code: String) -> Bool {
        code.count == 6 && code.unicodeScalars.allSatisfy { CharacterSet.decimalDigits.contains($0) }
    }

    static func authorizationValue(bootToken: String) -> String? {
        let token = bootToken.trimmingCharacters(in: .whitespacesAndNewlines)
        return token.isEmpty ? nil : "Bearer \(token)"
    }

    static func normalizedHost(_ host: String) -> String {
        host.trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    }

    static func workspaceURL(host: String) -> URL? {
        URL(string: normalizedHost(host) + "/api/mobile/workspace")
    }

    static func pairURL(host: String) -> URL? {
        URL(string: normalizedHost(host) + "/phone-link/v1/pair")
    }

    static func socketURL(host: String) -> URL? {
        guard let http = URL(string: normalizedHost(host)),
              var components = URLComponents(url: http, resolvingAgainstBaseURL: false) else { return nil }
        components.scheme = http.scheme?.lowercased() == "https" ? "wss" : "ws"
        components.path = "/phone-link/phone"
        return components.url
    }

    static func helloPayload(deviceToken: String) -> [String: Any] {
        ["v": 1, "type": "phone.hello", "deviceToken": deviceToken]
    }

    static func commandPayload(text: String, positionId: String, commandId: String) -> [String: Any] {
        [
            "v": 1,
            "type": "command.submit",
            "commandId": commandId,
            "text": text,
            "positionId": positionId,
        ]
    }

    static func parseMessage(_ text: String) -> [String: Any]? {
        guard let data = text.data(using: .utf8),
              let body = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              body["type"] as? String != nil else { return nil }
        return body
    }

    static func statusLabel(_ state: String) -> String {
        let labels = [
            "accepted": "电脑已接到",
            "running": "员工正在处理",
            "completed": "完成",
            "failed": "失败",
            "busy": "该员工正在忙",
            "needs_approval": "请在电脑上确认",
        ]
        return labels[state] ?? state
    }
}
